numba/npyufunc/ufuncbuilder.py

```diff
@@ -336,3 +362,82 @@ def build(self, cres):
                 ty = a
             dtypenums.append(as_dtype(ty).num)
         return dtypenums, ptr, env
+
+
+class GUFuncDispatcher(object):
+    """Holds the dispatch mechanics for gufuncs
+    """
+
+    def __init__(self, ufunc, sigs, impl_cache, typingctx, targetctx):
+        """ Initialiser for the GUFuncDispatcher
+        Arguments:
+         - ufunc the ufunc on which dispatch will take place
+         - sigs the signatures registered for dispatch
+         - impl_cache a map of signature to compile results
+         - typingctx the typing context
+         - targetctx the target context
+        """
+        self.ufunc = ufunc
+        self.sigs = sigs
+        self.impl_cache = impl_cache
+        self.typingctx = typingctx
+        self.targetctx = targetctx
+
+    def __call__(self, context, builder, sig, args):
+        """The lowering phase calls this to find an implementation, a suitable
+        function is found by matching the signature (sig) in `self.impl_cache`.
+        This is then staged and lowered.
+        """
+        cres = self.impl_cache[sig]
+        func_type = context.call_conv.get_function_type(sig.return_type,
+                                                        sig.args)
+        module = builder.module
+        entry_point = module.get_or_insert_function(
+            func_type, name=cres.fndesc.llvm_func_name)
+        entry_point.attributes.add("alwaysinline")
+
+        envptr = cres.environment.as_pointer(cres.target_context)
+
+        _, res = context.call_conv.call_function(
+            builder, entry_point, sig.return_type, sig.args, args,
+            env=envptr)
+        return res
+
+    def install_typing(self):
+        """Constructs and installs a typing class for the gufunc in the
+        typing context.
+        """
+        _ty_cls = type('GUFuncTyping_' + self.ufunc.__name__,
+                       (AbstractTemplate,),
+                       dict(key=self, generic=self._type_me))
+        self.typingctx.insert_user_function(self.ufunc, _ty_cls)
+
+    def _type_me(self, argtys, kwtys):
+        """Undertakes the typing of the gufunc against the signatures
+        for which it was compiled.
+        """
+        assert not kwtys
+        ufunc = self.ufunc
+        # create to-match sig of argtys ignoring data order
+        match_me = []
+        for argty in argtys:
+            if isinstance(argty, types.npytypes.Array):
+                match_me.append(argty.copy(layout='A'))
+            else:
+                match_me.append(argty)
+
+        sigmatcher = signature(types.void, *match_me)
+
+        if sigmatcher in self.sigs:
```

Comment:
The signature matching should use `ufunc_find_matching_loop ` (https://github.com/numba/numba/blob/c8127d8b16ccdb20d1c38c28a0045db85e560c59/numba/numpy_support.py#L288) or sth like that since numpy ufunc dispatch is different.